File: python3/maximum_difference_between_node_and_ancestor.py
from model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def maxAncestorDiff(self, root: TreeNode) -> int:
        if not root:
            return 0

        ans = 0
        stack = [(root, 10**6, -1)]

        while stack:
            node, minv, maxv = stack.pop()
            minv = min(minv, node.val)
            maxv = max(maxv, node.val)
            if not node.left and not node.right:
                ans = max(ans, abs(maxv - minv))
            if node.left:
                stack.append((node.left, minv, maxv))
            if node.right:
                stack.append((node.right, minv, maxv))

        return ans

File: golang/maximum_difference_between_node_and_ancestor.go
package main

import "golang/model"

/*
Space	: O(n)
Time	: O(n)
*/

func maxAncestorDiff(root *model.TreeNode) int {
	if root == nil {
		return 0
	}

	type Helper struct {
		Node *model.TreeNode
		Max  int
		Min  int
	}

	ans := 0
	stack := []Helper{
		Helper{
			Node: root,
			Max:  -1,
			Min:  100000,
		},
	}

	for len(stack) > 0 {
		temp := []Helper{}
		for _, helper := range stack {
			node := helper.Node
			maxv := Max(helper.Max, node.Val)
			minv := Min(helper.Min, node.Val)

			ans = Max(ans, Abs(maxv-minv))
			if node.Left != nil {
				temp = append(temp, Helper{
					Node: node.Left,
					Max:  maxv,
					Min:  minv,
				})
			}

			if node.Right != nil {
				temp = append(temp, Helper{
					Node: node.Right,
					Max:  maxv,
					Min:  minv,
				})
			}

		}
		stack = temp

	}

	return ans
}
